framework/src/headers/ip/v6/icmp/router_advertisement.rs

```diff
@@ -0,0 +1,336 @@
+        self.icmp.msg_type()
+    }
+
+    #[inline]
+    pub fn code(&self) -> u8 {
+        self.icmp.code()
+    }
+
+    #[inline]
+    pub fn checksum(&self) -> u16 {
+        self.icmp.checksum()
+    }
+
+    #[inline]
+    pub fn current_hop_limit(&self) -> u8 {
+        self.current_hop_limit
+    }
+
+    #[inline]
+    pub fn reserved_flags(&self) -> u8 {
+        self.reserved_flags
+    }
+
+    #[inline]
+    pub fn managed_addr_cfg(&self) -> bool {
+        get_bit(self.reserved_flags, MANAGED_CFG_ADDR_POS)
+    }
+
+    #[inline]
+    pub fn other_cfg(&self) -> bool {
+        get_bit(self.reserved_flags, OTHER_CFG_POS)
+    }
+
+    #[inline]
+    pub fn router_lifetime(&self) -> u16 {
+        u16::from_be(self.router_lifetime)
+    }
+
+    #[inline]
+    pub fn reachable_time(&self) -> u32 {
+        u32::from_be(self.reachable_time)
+    }
+
+    #[inline]
+    pub fn retrans_timer(&self) -> u32 {
+        u32::from_be(self.retrans_timer)
+    }
+}
+
+impl<T> IPv6Optionable for Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    fn parse_options(&self, payload_len: u16) -> HashMap<Icmpv6OptionType, Icmpv6Option> {
+        let mut options_map = HashMap::new();
+        unsafe {
+            let self_as_u8 = (self as *const Self) as *const u8;
+            let mut payload_offset = self.offset(); //track to make sure we don't go beyond v6 payload size
+
+            while payload_offset < (payload_len as usize) {
+                // start at beginning of the options
+                // the second byte is the header length in 8-octet unit excluding the first 8 octets
+                let seek_to = self_as_u8.offset(payload_offset as isize);
+
+                //lets get the first two indices which should be option_type and option_length fields
+                let option_meta = slice::from_raw_parts(seek_to, 2);
+
+                //Parse the option type field first then the option_length
+                let option_type: Option<Icmpv6OptionType> = FromPrimitive::from_u8(option_meta[0]);
+                let option_length = option_meta[1];
+                let option_length_octets = option_length * 8;
+
+                //Make sure the option is valid defined by the RFC. Note: a router can insert its own
+                //options defined outside of the spec.
+                if option_type.is_some() {
+                    //option_value_length is always total number of octets - 2(option type and length are always 2)
+                    let option_value_length = (option_length_octets - 2) as isize;
+
+                    let value_seek_to = self_as_u8.offset((payload_offset + 2) as isize);
+                    let option_value =
+                        slice::from_raw_parts(value_seek_to, option_value_length as usize);
+
+                    match option_type.unwrap() {
+                        Icmpv6OptionType::SourceLinkLayerAddress => options_map.insert(
+                            option_type.unwrap(),
```

Comment:
Not a fan of `if x.is_some() ... x.unwrap()`; Assuming there is a value there, it is safer to `for` over it...

```rust
  let x: Option<String> = Some("hi".to_string());
  for v in x {
    println!("Sup {}", v);
  }
```

Reply:
Also, we are unwrapping the option_type many times, pulling it from the option using the iterator would be easier to follow

Reply:
Note: final follow up here, we may want to log loudly if there is an option type that we do not support, so instead of a `for` in that case (which would only run if the option was `Some`) we would want to do a pattern match instead and have the `None` log loudly the unrecognized option type.

Reply:
I agree we should log loudly. I believe Z was working on the logging part although I can add that for Netbricks if no one is working on it. we have to call `is_some()` to make sure that its a valid option defined in our enum. We are parsing the options one by one and not at the same time. The logic goes  1) parse first byte...get option type 2) parse second byte, get option length 3) parse n bytes based on option length to get option_value. we repeat this process for N Options

Reply:
i fixed the multiple option unwrap BTW